Flow tracking (ver2)

PrasenjitAdhikary opened the pull request:
Proposed model for flows, with a new tree node `tag-metrics` for disaggregated stats:
This model is prepared based on review comments on PR https://github.com/open-traffic-generator/models-yang/pull/24 from [Greg Dennis](https://github.com/greg-dennis) 
```
module: open-traffic-generator-flow
  +--rw flows
     +--ro flow* [name]
        +--ro name           -> ../state/name
        +--ro state
        |  +--ro name?             string
        |  +--ro transmit?         boolean
        |  +--ro loss-pct?         otg-types:ieeefloat32
        |  +--ro out-frame-rate?   otg-types:ieeefloat32
        |  +--ro in-frame-rate?    otg-types:ieeefloat32
        |  +--ro out-rate?         otg-types:ieeefloat32
        |  +--ro in-rate?          otg-types:ieeefloat32
        |  +--ro counters
        |     +--ro in-octets?    otg-types:counter64
        |     +--ro in-pkts?      otg-types:counter64
        |     +--ro out-octets?   otg-types:counter64
        |     +--ro out-pkts?     otg-types:counter64
        +--ro tag-metrics
           +--ro tag-metric* [name-value]
              +--ro name-value    -> ../state/name-value
              +--ro state
                 +--ro name-value?   string
                 +--ro tags* []
                 |  +--ro tag-name?    string
                 |  +--ro tag-value
                 |     +--ro value-type?           enumeration
                 |     +--ro value-as-string?      string
                 |     +--ro value-as-hex?         otg-types:hex-string
                 |     +--ro value-as-bool?        boolean
                 |     +--ro value-as-counter64?   otg-types:counter64
                 |     +--ro value-as-float32?     otg-types:ieeefloat32
                 |     +--ro value-as-ipv4?        otg-types:ipv4-address
                 |     +--ro value-as-ipv6?        otg-types:ipv6-address
                 |     +--ro value-as-mac?         otg-types:mac-address
                 +--ro counters
                    +--ro in-octets?    otg-types:counter64
                    +--ro in-pkts?      otg-types:counter64
                    +--ro out-octets?   otg-types:counter64
                    +--ro out-pkts?     otg-types:counter64
```

If `metric_tags` are not present in the flow configuration (i.e. the user has not enabled tracking), gNMI response will contain empty value for `tag-metrics` node.


Assuming the following packet distribution for flow `F1`, some example gNMI query and results are described below:
![image](https://user-images.githubusercontent.com/79712606/230909845-5bfade8d-7e72-40f8-9cb9-cf62b157babe.png)
Here metric-tags are configured only for `VLAN ID` & `Src IP`, whereas `Dst IP` is shown to represent the packet and corresponding stats.

### 1. Query per-flow stats (including aggregated + disaggregated stats):
gNMI query path: “/flows/flow[name=F1]/”
gNMI response:
```
"updates": [
  {
	"Path": "flows/flow[name=f1]",
	"values": {
	  "flows/flow": {
		"open-traffic-generator-flow:name": "f1",
		"open-traffic-generator-flow:state": {                     // Contains the aggregated per-flow stats
		  "counters": {
			"in-octets": "2400",
			"in-pkts": "300",
			"out-octets": "2400",
			"out-pkts": "300"
		  },
		  "in-frame-rate": "QKAAAA==",
		  "name": "f1",
		  "out-frame-rate": "QCAAAA==",
		  "transmit": true
		},
		"open-traffic-generator-flow:tag-metrics": {              // Contains the disaggregated per-flow stats
		  "tag-metric": [
			{
			  "name-value": "ip=0x1&vlan=0x64",
			  "state": {
				"counters": {
				  "in-octets": "800",
				  "in-pkts": "100"
				},
				"name-value": "ip=0x1&vlan=0x64",
				"tags": [
				  {
					"tag-name": "ip",
					"tag-value": {
					  "value-as-hex": "0x1",
					  "value-type": "HEX"
					}
				  },
				  {
					"tag-name": "vlan",
					"tag-value": {
					  "value-as-hex": "0x64",
					  "value-type": "HEX"
					}
				  }
				]
			  }
			},
			{
			  "name-value": "ip=0x1&vlan=0xce",
			  "state": {
				"counters": {
				  "in-octets": "160",
				  "in-pkts": "20"
				},
				"name-value": "ip=0x1&vlan=0xce"
				"tags": [
				  {
					"tag-name": "ip",
					"tag-value": {
					  "value-as-hex": "0x1",
					  "value-type": "HEX"
					}
				  },
				  {
					"tag-name": "vlan",
					"tag-value": {
					  "value-as-hex": "0xce",
					  "value-type": "HEX"
					}
				  }
				]
			  }
			},
			{
			  "name-value": "ip=0x2&vlan=0x64",
			  "state": {
				"counters": {
				  "in-octets": "1200",
				  "in-pkts": "150"
				},
				"name-value": "ip=0x2&vlan=0x64"
				"tags": [
				  {
					"tag-name": "ip",
					"tag-value": {
					  "value-as-hex": "0x2",
					  "value-type": "HEX"
					}
				  },
				  {
					"tag-name": "vlan",
					"tag-value": {
					  "value-as-hex": "0x64",
					  "value-type": "HEX"
					}
				  }
				]
			  }
			},
			{
			  "name-value": "ip=0x2&vlan=0xce",
			  "state": {
				"counters": {
				  "in-octets": "240",
				  "in-pkts": "30"
				},
				"name-value": "ip=0x2&vlan=0xce"
				"tags": [
				  {
					"tag-name": "ip",
					"tag-value": {
					  "value-as-hex": "0x2",
					  "value-type": "HEX"
					}
				  },
				  {
					"tag-name": "vlan",
					"tag-value": {
					  "value-as-hex": "0xce",
					  "value-type": "HEX"
					}
				  }
				]
			  }
			}
		  ]
		}
	  }
	}
  }
]

```

**Note:** Assumption, Metric Tag for Src IP is configured only for the last 8 bits. Hence IP Src address is matched for the last 8 bits appearing in the result. 

### 2. Query only the disaggregated stats (based on all metric tags present in the configuration):
gNMI query path: “/flows/flow[name=F1]/tag-metrics”
gNMI response:

```
"updates": [
  {
	"Path": "flows/flow[name=f1]/tag-metrics/tag-metric[name-value=ip=0x1&vlan=0xce]",
	"values": {
	  "flows/flow/tag-metrics/tag-metric": {
		"open-traffic-generator-flow:name-value": "ip=0x1&vlan=0xce",
		"open-traffic-generator-flow:state": {
		  "counters": {
			"in-octets": "160",
			"in-pkts": "20"
		  },
		  "name-value": "ip=0x1&vlan=0xce"
		  "tags": [
		    {
			  "tag-name": "ip",
			  "tag-value": {
			    "value-as-hex": "0x1",
			    "value-type": "HEX"
			  }
		    },
		    {
			  "tag-name": "vlan",
			  "tag-value": {
			    "value-as-hex": "0xce",
			    "value-type": "HEX"
			  }
		    }
		  ]
		}
	  }
	}
  },
  {
	"Path": "flows/flow[name=f1]/tag-metrics/tag-metric[name-value=ip=0x2&vlan=0xce]",
	"values": {
	  "flows/flow/tag-metrics/tag-metric": {
		"open-traffic-generator-flow:name-value": "ip=0x2&vlan=0xce",
		"open-traffic-generator-flow:state": {
		  "counters": {
			"in-octets": "240",
			"in-pkts": "30"
		  },
		  "name-value": "ip=0x2&vlan=0xce"
		  "tags": [
		    {
			  "tag-name": "ip",
			  "tag-value": {
			    "value-as-hex": "0x2",
			    "value-type": "HEX"
			  }
		    },
		    {
			  "tag-name": "vlan",
			  "tag-value": {
			    "value-as-hex": "0xce",
			    "value-type": "HEX"
			  }
		    }
		  ]
		}
	  }
	}
  },
  {
	"Path": "flows/flow[name=f1]/tag-metrics/tag-metric[name-value=ip=0x1&vlan=0x64]",
	"values": {
	  "flows/flow/tag-metrics/tag-metric": {
		"open-traffic-generator-flow:name-value": "ip=0x1&vlan=0x64",
		"open-traffic-generator-flow:state": {
		  "counters": {
			"in-octets": "800",
			"in-pkts": "100"
		  },
		  "name-value": "ip=0x1&vlan=0x64"
		  "tags": [
		    {
			  "tag-name": "ip",
			  "tag-value": {
			    "value-as-hex": "0x1",
			    "value-type": "HEX"
			  }
		    },
		    {
			  "tag-name": "vlan",
			  "tag-value": {
			    "value-as-hex": "0x64",
			    "value-type": "HEX"
			  }
		    }
		  ]
		}
	  }
	}
  },
  {
	"Path": "flows/flow[name=f1]/tag-metrics/tag-metric[name-value=ip=0x2&vlan=0x64]",
	"values": {
	  "flows/flow/tag-metrics/tag-metric": {
		"open-traffic-generator-flow:name-value": "ip=0x2&vlan=0x64",
		"open-traffic-generator-flow:state": {
		  "counters": {
			"in-octets": "1200",
			"in-pkts": "150"
		  },
		  "name-value": "ip=0x2&vlan=0x64"
		  "tags": [
		    {
			  "tag-name": "ip",
			  "tag-value": {
			    "value-as-hex": "0x2",
			    "value-type": "HEX"
			  }
		    },
		    {
			  "tag-name": "vlan",
			  "tag-value": {
			    "value-as-hex": "0x64",
			    "value-type": "HEX"
			  }
		    }
		  ]
		}
	  }
	}
  }
]
            
```

### 3. Query disaggregated stats for VLAN ID = 100 & Src IP = 10.10.10.1 (multi-field tracking with filtering):
gNMI query path: "flows/flow[name=f1]/tag-metrics/tag-metric[name-value='ip=0x1&vlan=0x64']"
gNMI response:
```
"updates": [
  {
	"Path": "flows/flow[name=f1]/tag-metrics/tag-metric[name-value=ip=0x1&vlan=0x64]",
	"values": {
	  "flows/flow/tag-metrics/tag-metric": {
		"open-traffic-generator-flow:name-value": "ip=0x1&vlan=0x64",
		"open-traffic-generator-flow:state": {
		  "counters": {
			"in-octets": "800",
			"in-pkts": "100"
		  },
		  "name-value": "ip=0x1&vlan=0xce"
		  "tags": [
		    {
			  "tag-name": "ip",
			  "tag-value": {
			    "value-as-hex": "0x1",
			    "value-type": "HEX"
			  }
		    },
		    {
			  "tag-name": "vlan",
			  "tag-value": {
			    "value-as-hex": "0xce",
			    "value-type": "HEX"
			  }
		    }
		  ]

		}
	  }
	}
  }
]

```

### 4. Query only the aggregated stats:
**Option A:** Query only for the "state" 
gNMI query path: "flows/flow[name=f1]/state]"
gNMI response:
```
"updates": [
  {
	"Path": "flows/flow[name=f1]/state/counters",
	"values": {
	  "flows/flow/state/counters": {
		"open-traffic-generator-flow:in-octets": "2000",
		"open-traffic-generator-flow:in-pkts": "200",
		"open-traffic-generator-flow:out-octets": "1000",
		"open-traffic-generator-flow:out-pkts": "59"
	  }
	}
  },
  {
	"Path": "flows/flow[name=f1]/state/in-frame-rate",
	"values": {
	  "flows/flow/state/in-frame-rate": "QKAAAA=="
	}
  },
  {
	"Path": "flows/flow[name=f1]/state/name",
	"values": {
	  "flows/flow/state/name": "f1"
	}
  },
  {
	"Path": "flows/flow[name=f1]/state/out-frame-rate",
	"values": {
	  "flows/flow/state/out-frame-rate": "QCAAAA=="
	}
  },
  {
	"Path": "flows/flow[name=f1]/state/transmit",
	"values": {
	  "flows/flow/state/transmit": true
	}
  }
]


```

**Option B:** Remove 'tag_metrics' from the configuration and then do a fresh query: 
gNMI query path: "flows/flow[name=f1]/]"
gNMI response:
```
"updates": [
  {
	"Path": "flows/flow[name=f1]",
	"values": {
	  "flows/flow": {
		"open-traffic-generator-flow:name": "f1",
		"open-traffic-generator-flow:state": {
		  "counters": {
			"in-octets": "2000",
			"in-pkts": "200",
			"out-octets": "1000",
			"out-pkts": "81"
		  },
		  "in-frame-rate": "QKAAAA==",
		  "name": "f1",
		  "out-frame-rate": "QCAAAA==",
		  "transmit": true
		}
	  }
	}
  }
]

```

anjan-keysight reviewed (approved):
Good to go.

alakendu commented:
I am ok with the `tag-metrics` .  
A query related to overall `flows/flow`. Model do not have `src-port` and `dst-port`. In our current implementation this is fine as we are supporting single endpoint. But I feel those will require when we will expose for multiple endpoints situated in different `Rx Ports`

greg-dennis commented:
The current proposal requires encoding of the key value to do a specific lookup, but it also could require _decoding_ of the key values when doing a general scan of the table. One way to avoid any need for decoding is to add a child of tag_metrics/state for that is a container for all the key-value pairs. The key-value pairs would not be a generic string-string mapping; instead the keys could be specifically enumerated to the allowed tags and strongly-typed to their value (string, int, bool, etc). That is, you could get a gnmi response like:

```
{
	"name-value": "ip=0x1&vlan=0x64",
	"state": {
	  	"tags": {
			"ip": "0x1",
			"vlan": "0x64",
		},
		"counters": {
			"in-octets": "800",
			"in-pkts": "100"
		},
		"name-value": "ip=0x1&vlan=0x64",
	}
},
{
	"name-value": "ip=0x1&vlan=0xce",
	"state": {
		"tags": {
			"ip": "0x1",
			"vlan": "0xce",
		},
		"counters": {
			"in-octets": "160",
			"in-pkts": "20"
		},
		"name-value": "ip=0x1&vlan=0xce",
	}
}
```

Any thoughts on that?

To further eliminate the need for encoding, we could take that response and just replace the keys with row index numbers instead. (Technically, we could make the list unkeyed, but unkeyed lists don't play well with all ygot/ygnmi utilities.) Using a row index key, the response would look like:

```
{
	"row": "0",
	"state": {
	  	"tags": {
			"ip": "0x1",
			"vlan": "0x64",
		},
		"counters": {
			"in-octets": "800",
			"in-pkts": "100"
		},
		"row": "0"
	}
},
{
	"row": "1",
	"state": {
		"tags": {
			"ip": "0x1",
			"vlan": "0xce",
		},
		"counters": {
			"in-octets": "160",
			"in-pkts": "20"
		},
		"row": "1"
	}
}
```

One negative of this approach is that looking up the metric for a specific node-value combination is now an O(N) operation instead of O(1). The question is what would be the meaningful performance hit of doing this in practice. If it's negligible, it might be worth it to avoid the encoding. If it could be substantial, might be reason to stick with the encoding.

If we're unsure, we could potentially go with this latter approach and should be find a performance problem, change the model to make the key an encoded string. While that would technically be a breaking change, it seems unlikely most users are querying or accessing the row index for anything meaningful, so the breaking change might be minimal.

To be clear, I'm not sold on this latter approach -- I'd like your opinion.

PrasenjitAdhikary commented:
@greg-dennis Our proposal in response to https://github.com/open-traffic-generator/models-yang/pull/27#issuecomment-1507457503

In the proposed OTG model, the user can set tracking on any packet header field, and track one/more fields at a time, on ingress/egress/both sides (ref PR https://github.com/open-traffic-generator/models/pull/312).

To match with this OTG model, we came up with this concept of encoding, so that the gNMI yang model can generically support the tracking, for all header fields present today in the model as well as to be added in the future, at the cost of parsing (decoding/encoding).
If we try to create an enum list for all trackable fields, the list is difficult to manage for the following reason:  
	- to include all possible trackable fields from all supported existing headers. Also, need to be enhanced every time a new trackable field is added to the OTG model.  
	- the name should be unique (if two protocol field has the same field name, need some decoration to distinguish between the fields. Example: ipv4_src, ipv6_src, ipv6_src, ipv6_dst)  
	- Since multiple fields can be tracked and that can be done on the same protocol, need to distinguish between the headers. Considering Q-In-Q, IP-in-IP, and multiple MPLS headers, we need to create a redundant set of enums like:  

		vlan_id.1
		vlan_id.2
		...
		vlan_id.M
		mpls_label.1
		mpls_label.2
		...
		mpls_label.N
		ipv4_src.inner
		ipv4_src.outer
		// limit of M, N to be decided

And each enum field can be associated with two strongly typed values. In the OTG model, the user can track either the whole field or some bits of that field. For example, the ipv4_src field can be tracked for the last 8 bits (suitable mostly for egress tracking) or for the entire field (suitable for ingress tracking). If the entire field is tracked, the matched value will be shown in the native format of the field (here IPv4). If some bits are tracked, the matched value will be shown in hex format. So, for the same ipv4_src field, the value could be "1.1.1.1" or 0x12, depending on how tracking has been set on the field. This can effectively double the enum number or need to split the value structure for each enum with two sub-fields (hex and the native value)
	
As part of the delivery, we will add some helper functions, for encoding/decoding, so that the gNMI results can be shown nicely in a tabular manner. If this approach is accepted, we can raise a PR for including the helper function in the ondatra library.

Alternatively, we can try the following yang-model, which can solve the issue by (a) an unkeyed list for tags - so that each tag has a separate structure having name & value (b) the value is strongly typed using a key list where the key is an enumerated data type.  

```
Yang model (only for the tag-metrics part):
+--ro tag-metrics
   +--ro tag-metric* [name-value]
	  +--ro name-value    -> ../state/name-value
	  +--ro state
		 +--ro name-value?        string
		 +--ro tags*
		 +  +--ro tag-name?       string
		 +  +--ro tag-value* [value-type]
		 +     +--ro value-type    -> ../state/value-type
		 +     +--ro state
		 +        +--ro value-type?       enumeration
		 +        +--ro value-as-string?  string
		 +        +--ro value-as-hex?     otg-types:hex-string
		 +        +--ro value-as-bool?    boolean
		 +        +--ro value-as-uint64?  otg-types:counter64
		 +        +--ro value-as-float?   otg-types:ieeefloat32
		 +        +--ro value-as-mac?     otg-types:mac-address
		 +        +--ro value-as-ipv4?    otg-types:ipv4-address
		 +        +--ro value-as-ipv6?    otg-types:ipv6-address
		 +--ro counters
			+--ro in-octets?    otg-types:counter64
			+--ro in-pkts?      otg-types:counter64
			+--ro out-octets?   otg-types:counter64
			+--ro out-pkts?     otg-types:counter64
```
Example result with **Option 1** (encoded name-value is still present, but the gNMI client can ignore this value and process only the tags section):
```
{
	"name-value": "ip=10.10.10.1&vlan=0x64",
	"state": {
		"name-value": "ip=10.10.10.1&vlan=0x64",
	  	"tags": [
			{
				"tag-name": "ip",
				"tag-value" [{
					"value-type": "ipv4"
					"state" {
						"value-type": "ipv4"
						"value-as-ipv4" : "10.10.10.1"
					}				
				}]
			},
			{
				"tag-name": "vlan",
				"tag-value" [{
					"value-type": "hex"
					"state" {
						"value-type": "hex"
						"value-as-hex" : "0x64"
					}				
				}]
			}
		],
		"counters": {
			"in-octets": "800",
			"in-pkts": "100"
		},
	}
},
{
	"name-value": "ip=10.10.10.1&vlan=0xce",
	"state": {
		"name-value": "ip=10.10.10.1&vlan=0xce",
	  	"tags": [
			{
				"tag-name": "ip",
				"tag-value" [{
					"value-type": "ipv4"
					"state" {
						"value-type": "ipv4"
						"value-as-ipv4" : "10.10.10.1"
					}				
				}]
			},
			{
				"tag-name": "vlan",
				"tag-value" [{
					"value-type": "hex"
					"state" {
						"value-type": "hex"
						"value-as-hex" : "0xce"
					}				
				}]
			}
		],
		"counters": {
			"in-octets": "160",
			"in-pkts": "20"
		},
	}
},
```
Example result with **Option 2** (encoded name-value is replaced with a row number (some autogenerated value):
```
{
	"row": "0",
	"state": {
		"row": "0",
	  	"tags": [
			{
				"tag-name": "ip",
				"tag-value" [{
					"value-type": "ipv4"
					"state" {
						"value-type": "ipv4"
						"value-as-ipv4" : "10.10.10.1"
					}				
				}]
			},
			{
				"tag-name": "vlan",
				"tag-value" [{
					"value-type": "hex"
					"state" {
						"value-type": "hex"
						"value-as-hex" : "0x64"
					}				
				}]
			}
		],
		"counters": {
			"in-octets": "800",
			"in-pkts": "100"
		},
	}
},
{
	"row": "1",
	"state": {
		"row": "1",
	  	"tags": [
			{
				"tag-name": "ip",
				"tag-value" [{
					"value-type": "ipv4"
					"state" {
						"value-type": "ipv4"
						"value-as-ipv4" : "10.10.10.1"
					}				
				}]
			},
			{
				"tag-name": "vlan",
				"tag-value" [{
					"value-type": "hex"
					"state" {
						"value-type": "hex"
						"value-as-hex" : "0xce"
					}				
				}]
			}
		],
		"counters": {
			"in-octets": "160",
			"in-pkts": "20"
		},
	}
},
```

Some advantages of option 1:
	- if the gNMI client is fetching all the drill-down metrics, both options are the same. The client has to process the entire result set, and in that case, the client library would always process only "tags" (and ignore "name-value" - applicable for option 1)
	- if the client is interested to retrieve metrics only for specific tags, then encoded "name-value" can be formed (using some helper function) and used as the key in the gNMI query path, using option 1. This support will not be available for option 2, where the key for each drill-down metric is autogenerated number, which is not known to the client in advance.

greg-dennis commented:
Thanks for the detailed response.

One clarification: although I used the word "enumerated," I wasn't suggesting any new enum types.

One thing I believe I just realized now ... the keys like "ip" and "vlan" in your example are the names the _user_ chose in their configuration, correct? My mind went to the IxNetwork model in which the tracking usually uses preset names. Your model is much more flexible and makes sense. My suggestion of having a field for each possible key doesn't make sense in this model -- got it.

Another question. Looking at this:

```
"tag-value" [{
	"value-type": "ipv4"
	"state" {
		"value-type": "ipv4"
		"value-as-ipv4" : "10.10.10.1"
	}				
}]
```

Am I to take the brackets to mean that a single tag may have _multiple_ values for a single combination/row? Why would that be?

greg-dennis commented:
That `tag-value` looks great to us.

We had a further discussion about whether we want to stick with the encoded key (aided by a helper library) or just have row numbers. Right now, we think there is probably enough reason to go with the encoded key.

One question about the encoded key: what order do the keys have to appear in? The two alternatives we thought of were:

1. configuration order, i.e. the order in which the tags are added to MetricTags
2. alphabetical order of the tag name

Alphabetical order had the advantage that the helper function could take a simple Go map and ensure the resulting encoded key is in the right order. With configuration order, the user would have to supply either (a) a list of structs or (b) a map _and_ the MetricTags object to ensure the are in the right order. If (a) is done, it could break under a refactoring that moves lines of code around. Might be reason enough to use alphabetical order?

PrasenjitAdhikary commented:
> That `tag-value` looks great to us.
> 
> We had a further discussion about whether we want to stick with the encoded key (aided by a helper library) or just have row numbers. Right now, we think there is probably enough reason to go with the encoded key. To be clear, I mean use the encoded key _in addition to_ the tags list.
> 
> One question about the encoded key: what order do the keys have to appear in? The two alternatives we thought of were:
> 
> 1. configuration order, i.e. the order in which the tags are added to MetricTags
> 2. alphabetical order of the tag name
> 
> Alphabetical order had the advantage that the helper function could take a simple Go map and ensure the resulting encoded key is in the right order. With configuration order, the user would have to supply either (a) a list of structs or (b) a map _and_ the MetricTags object to ensure the are in the right order. If (a) is done, it could break under a refactoring that moves lines of code around. Might be reason enough to use alphabetical order?

I have updated the model and PR description/examples to reflect the encoded key along with the tags list – based on the discussion till now.
In the encoded key, the order of the tags will be in alphabetical order of the tag name. I have added this condition in the model description, with examples.

ashutshkumr commented on artifacts/open-traffic-generator-flow.txt:
```diff
@@ -1,17 +1,39 @@
+        |     +--ro in-pkts?      otg-types:counter64
+        |     +--ro out-octets?   otg-types:counter64
+        |     +--ro out-pkts?     otg-types:counter64
+        +--ro tag-metrics
```
this needs to be renamed to `tagged-metrics` or `tagged-metric` wherever applicable

ashutshkumr commented on artifacts/open-traffic-generator-flow.txt:
```diff
@@ -1,17 +1,39 @@
+        |     +--ro out-pkts?     otg-types:counter64
+        +--ro tag-metrics
+           +--ro tag-metric* [name-value]
+              +--ro name-value    -> ../state/name-value
```
since it's a query string (https://en.wikipedia.org/wiki/Query_string), it should called either one of following:
1. query-string
2. tag-filter
3. name-value-pairs

PrasenjitAdhikary replied:
We can not call it just 'query-string'. It is one 'key' (or identifier), to identify a `tag-metric` result, which can be used to form the gNMI query - at the same time, this key is present in the response also. So, `query-string` or `tag-filter` is not appropriate. A longer name could be 'tags-as-name-value-pairs-string' or any appropriate shorter name. 'name-value-pairs' could be reasonable.

ashutshkumr commented on artifacts/open-traffic-generator-flow.txt:
```diff
@@ -1,17 +1,39 @@
+                 +--ro tags* []
+                 |  +--ro tag-name?    string
+                 |  +--ro tag-value
+                 |     +--ro value-type?           enumeration
+                 |     +--ro value-as-string?      string
+                 |     +--ro value-as-hex?         otg-types:hex-string
+                 |     +--ro value-as-bool?        boolean
+                 |     +--ro value-as-counter64?   otg-types:counter64
+                 |     +--ro value-as-float32?     otg-types:ieeefloat32
+                 |     +--ro value-as-ipv4?        otg-types:ipv4-address
+                 |     +--ro value-as-ipv6?        otg-types:ipv6-address
+                 |     +--ro value-as-mac?         otg-types:mac-address
```
Do we need to prefix all of them with `value` or `value-as`, since parent is already called `tag-value` ?

greg-dennis replied:
After further discussion, we prefer to leave it as separate fields. I don't have a strong preference on the "value-as-" prefix. I could see a case for dropping the "-as-" part of each name. I'll defer to you.

greg-dennis replied:
On second thought, the "-as-" seems to add some needed clarity. Carry on
